xarray/backends/netCDF4_.py

```diff
@@ -25,14 +26,29 @@
                   '|': 'native'}
 
 
+class VariableWrapper(object):
+    """ Ensures that variable attributes and data access operates on open datastore """
+    def __init__(self, datastore, variable_name):
+        self.var = variable_name
+        self.datastore = datastore
+
+    def __getattr__(self, attr):
+        with self.datastore.ensure_open():
+            return getattr(self.datastore.ds.variables[self.var], attr)
+
+    def __getitem__(self, key):
+        with self.datastore.ensure_open():
+            return operator.getitem(self.datastore.ds.variables[self.var], key)
+
+
 class BaseNetCDF4Array(NDArrayMixin):
     def __init__(self, variable_name, datastore):
         self.datastore = datastore
         self.variable_name = variable_name
 
     @property
     def array(self):
-        return self.datastore.ds.variables[self.variable_name]
+        return VariableWrapper(self.datastore, self.variable_name)
```

Comment:
Interesting- it looks like `__array__` above and `self.dtype` weren't necessarily needed.  Tests pass at least.  Should I be concerned about this issue?  For instance, perhaps they will be needed for `h5netcdf`.  I don't fully know the xarray internal structure well enough to understand the significance here, but am curious to find out if it is easy to explain.